### ISSUE TITLE
[Example VPN moduile] Small module for VPN, works great

### ISSUE THREAD

**Issue by mikedebian** [Alexays/Waybar#3583]
.config/waybar/config.jsonc`

`....`
`"modules-right": [`
`....`
`"custom/vpn", `

The module

`....`
`"custom/vpn": {`
`"interval": 3,`
`"format": "{}",`
`"exec": "ip add show | grep -qF tun0 && echo 🔒 Connected || echo 🔓 Disconnected",`
`"max-length": "100",`
`"on-click": "nmcli connection up insertovpnfilenamewithoutextension",`
`"on-click-right": "nmcli connection down insertovpnfilenamewithoutextension"`
`},`
`.....`

The style.css

` ....`
`#custom-vpn `
`      font-family: "JetBrainsMono Nerd Font";`
`      font-size: 24px;`
`      padding-left: 2px;`
`      padding-right: 2px;`
`      transition: all 0.25s cubic-bezier(0.165, 0.84, 0.44, 1);`
`}`
`....`

Sorry for the "markdown", but I know that you know how it is supposed to be properly formatted even though I'm unable to do so here.. Lack of skills on my part.

This shows a default state of disconnected, with the open padlock character from the font (It shows up as an icon in this example though. Again, lack of markdown skils. Just search for the character "padlock" within the font. Copy pasting this will also work).

Every interval of 3 seconds, a command gets executed that pulls the status from the tun0 interface, if it is disconnected or connected, and it outputs it through the "format": "{}.

On-click sets up the openvpn connection, and when the exec command pulls the update, if the openvpn connection from On-click was succesful, it will change from disconnected to connected. Right click terminates the connection, wait 3 seconds and it will pull the status from tun0 and show disconnected.

I just like the styling. Modify anything as you wish.

I think it rocks because the status of the module will change on its own without user input. If the connection dies, it will automatically go from connected to disconnected. If I start the connection from the terminal, it will still pull the status on the tun0 interface and toggle between connected or disconnected.

If I had set the vpn connection in an autostart file like ~/.config/labwc/autostart , the button would have showed a locked padlock and the words Connected would have been the default state. I also wanted to use common tools, instead of for example clients like bluetit, hummingbird,goldcrest, etc. so other people could use it too. Sharing is caring.

**Comment by mikedebian** [comment]
![swappy-20240908_212745](https://github.com/user-attachments/assets/5d619d67-db5b-4c8a-9fed-65edb31e9d98)
![swappy-20240908_212804](https://github.com/user-attachments/assets/8d68b972-1bb1-4fe3-b9eb-6b04b606625a)

**Comment by cferg296** [comment]
@mikedebian I have not tried this yet and wont be able to until tomorrow. If this works then you will be my hero. Ive been trying to make a module for displaying VPN connections for a while and haven't figured it our or found a solution that worked. (i use mullvad)

**Comment by mikedebian** [comment]
> @mikedebian I have not tried this yet and wont be able to until tomorrow. If this works then you will be my hero. Ive been trying to make a module for displaying VPN connections for a while and haven't figured it our or found a solution that worked. (i use mullvad)

So cool! Hey, you might have to adapt it to what you use, for example the interface might not be tun0. Please provide feedback whether it works or not :) I really want this to be as standardized as possible.

**Comment by cferg296** [comment]
@mikedebian Doesnt work for me. All i did though was interchange with mullvad's name but doesnt work. You mentioned tun0 might not work but im not sure what tun0 is or what i could do to further adapt it. Any ideas? 
![2024-09-10-234440_hyprshot](https://github.com/user-attachments/assets/ebe5d72d-96ae-44fc-bbbb-681674581541)

**Comment by cferg296** [comment]
Actually wait nevermind. It DOES work (sort of). It does display on my waybar, however clicking on it does not connect or disconnect from the VPN. Also it will say "disconnected" even when i am connected to my vpn 
![image](https://github.com/user-attachments/assets/b282226e-a029-4fd4-b0f1-1d0b00d5a9ac)

**Comment by mikedebian** [comment]
Well, I have no experience using mullvad! Is your connection stored in ~/.cert/nm-openvpn ? Something like

'AirVPN_Grus_UDP-443-Entry3-ca.pem'
'AirVPN_Grus_UDP-443-Entry3-cert.pem'
'AirVPN_Grus_UDP-443-Entry3-key.pem'
'AirVPN_UDP-443-Entry3-tls-crypt.pem'

This is where networkmanager looks for the files.

With the VPN on, what is the output of
'ip add show'

**Comment by cferg296** [comment]
@mikedebian I dont have a ~/.cert folder.

And i dont want to show the output as it displays an IP address

**Comment by cferg296** [comment]
Wait i have good news. I WAS able to get it to at least display whether im connected or not. I switched the setting from wireguard to openvpn and now it will properly display it. However i still cant figure out how to get it to connect or disconnect based on clicking

**Comment by cferg296** [comment]
@mikedebian  wait i figured it out! I changed it to this and now everything works perfectly! So currently your setup only works if a vpn is using openvpn and not wireguard. And in order to get it to connect with mullvad you just use "mullvad connect/disconnect". But it does work. Thank you!
![image](https://github.com/user-attachments/assets/348061f6-8b3b-4b5d-b5b4-e9c8b7986806)

**Comment by mikedebian** [comment]
Hey that was my next question, if you were using WireGuard or OpenVPN :) Very cool, another success!

**Comment by cferg296** [comment]
Let me know if you find a way to get it to work with wireguard.

**Comment by mikedebian** [comment]
> Let me know if you find a way to get it to work with wireguard.

So, I don't use wireguard although I did some years ago...

I've been reading the guide at https://mullvad.net/en/help/wireguard-and-mullvad-vpn and under section 4. and 5. it states.

`4. Connect with WireGuard`
``
Log in to the root account with su --login and change directory using cd /etc/wireguard. Then run the command below but  replace se-mma-wg-001 with the server config file that you want to use.
``
`wg-quick up se-mma-wg-001`
``
5. Verify your connection
``
`Check if you are connected to Mullvad:`
`curl https://am.i.mullvad.net/connected`

This seems on par with the OpenVPN guide that mullvad provides, although the command is different.

`wg-quick up se-mma-wg-001` instead of
'nmcli connection up se-mma-wg-001'

So, check to see which interface wireguard uses with 'ip add show' and modify the script accordingly. 

Let me know if you need some help :)

**Comment by mikedebian** [comment]
I've never done a pull request before, so maybe I can tag you on this one @Alexays  ?

**Comment by mikedebian** [comment]
> Let me know if you find a way to get it to work with wireguard.

@cferg296 Were you able to make it work?

**Comment by cferg296** [comment]
> > Let me know if you find a way to get it to work with wireguard.
> 
> @cferg296 Were you able to make it work?

Sorry its been a while! So i havnt tried your fix with wireguard but i will in a bit, but i wanted to say that how it is DOES allow mullvad wireguard to connect and disconnect, however the actual symbol that is displayed will not change no matter if you are connected or not.

**Comment by cferg296** [comment]
> > Let me know if you find a way to get it to work with wireguard.
> 
> @cferg296 Were you able to make it work?

You are right that if its wg0 then it connects via wineguard and if it says tun0 then it works with openvpn. I am now trying to figure out how to have it work regardless of which one you are using. I thought having it say tun0 || wg0 would work but it doesnt.

**Comment by Aa123123-aa** [comment]
简单设置一键翻墙，速度瞬间提升5～10倍。我在大陆地区如何上网，从事涉外网络配置专业工程师10年经验分享。不走弯路，不被坑！2024年度最具性价比VPN机场真心推荐。下载地址：
https://expvpn.net
https://vpndapps.com
https://letsvpnt.com
https://letskuailian.com

**Comment by RodsonMatos** [comment]
I did this in case you're using mullvad through their package:
(jq is a dependency)
```
"custom/mullvadvpn": {
       "interval": 1,
       "format": "{}",
       "exec": "mullvad status -j | jq -r '.details.location.hostname' | grep -qF wg && echo 🔒 $(mullvad status -j | jq -r '.details.location.hostname') || echo 🔓 Disconnected", // shows the relay if connected
       "max-length": "100",
       "on-click": "mullvad connect",
       "on-click-right": "mullvad disconnect",
       "on-click-middle": "mullvad-vpn",
       "tooltip-format": "Left click: connect\nRight click: disconnect\nMiddle click: open gui",
    },
```

**Comment by Alexays** [comment]
Closing — this is a config showcase rather than a bug or feature request; it's a better fit for GitHub Discussions. Thanks for sharing it! 🙏
